pkg/controller/sls/base.go

```diff
@@ -0,0 +1,138 @@
+/*
+
+ Copyright 2021 The Crossplane Authors.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+
+*/
+
+package sls
+
+import (
+	sdk "github.com/aliyun/aliyun-log-go-sdk"
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	"github.com/crossplane/crossplane-runtime/pkg/event"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+	"github.com/pkg/errors"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	"github.com/crossplane/provider-alibaba/apis/sls/v1alpha1"
+	slsclient "github.com/crossplane/provider-alibaba/pkg/clients/sls"
+)
+
+// BaseObserve is the common logic for controller Observe reconciling
+func BaseObserve(mg resource.Managed, c slsclient.LogClientInterface) (managed.ExternalObservation, error) {
+	cr, ok := mg.(*v1alpha1.Project)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotProject)
+	}
+	projectName := cr.Spec.ForProvider.ProjectName
+	project, err := c.Describe(projectName)
+	if slsclient.IsNotFoundError(err) {
+		return managed.ExternalObservation{
+			ResourceExists: false,
+		}, nil
+	}
+
+	if err != nil {
+		return managed.ExternalObservation{}, err
+	}
+
+	cr.Status.AtProvider = slsclient.GenerateObservation(project)
+	var upToDate bool
+	if (projectName == project.Name) && (cr.Spec.ForProvider.Description == project.Description) {
+		upToDate = true
+		cr.SetConditions(xpv1.Available())
+	}
+
+	return managed.ExternalObservation{
+		ResourceExists:    true,
+		ResourceUpToDate:  upToDate,
+		ConnectionDetails: getConnectionDetails(project),
+	}, nil
+}
+
+// BaseCreate is the logic for Create reconciling
+func BaseCreate(mg resource.Managed, c slsclient.LogClientInterface) (managed.ExternalCreation, error) {
+	cr, ok := mg.(*v1alpha1.Project)
+	if !ok {
+		return managed.ExternalCreation{}, errors.New(errNotProject)
+	}
+	name := cr.Spec.ForProvider.ProjectName
```

Comment:
We should remove `cr.Spec.ForProvider.ProjectName` and instead use `meta.GetExternalName(cr)`. It turns out this pattern is very poorly documented, but we require it from all resources these days. You can see an example in the link below.

The `meta.GetExternalName()` should be used anywhere you currently use `cr.Spec.ForProvider.ProjectName`. function will return the value of the `crossplane.io/external-name` annotation, which Crossplane ensures is always set. If the user does not specify the annotation it is set automatically to the `metadata.name` of the `Project` managed resource. This allows Crossplane users to have consistent control over the "actual" or "external" name of their managed resources, rather than having to figure out which spec field corresponds to the name.

https://github.com/crossplane/provider-alibaba/blob/0cfdaf72d/pkg/controller/database/rdsinstance.go#L217